CalendarApp/Logic Handlers/EventSyncHandler.m

```diff
@@ -0,0 +1,146 @@
+        [self.userData setObject:[NSDate date] forKey:@"lastUpdated"];
+        [self.userData synchronize];
+        return;
+    }
+    [self.parseEventHandler queryEventsAfterUpdateDate:[self.userData objectForKey:@"lastUpdated"]
+                                            completion:^(BOOL success,
+                                                         NSMutableArray<Event *> * _Nullable events,
+                                                         NSDate * _Nullable updatedDate,
+                                                         NSString * _Nullable error) {
+        if (!success) {
+            // TODO: Error handling
+        } else {
+            NSArray<LocalChange *> *localChanges = [self.context executeFetchRequest:LocalChange.fetchRequest error:nil];
+            
+            SyncConflictHandler *conflictHandler = [[SyncConflictHandler alloc] init];
+            NSArray<LocalChange *> *keptChanges = [conflictHandler getChangesToSyncWithOnlineEvents:events offlineChanges:localChanges];
+            [self syncLocalChanges:keptChanges];
+            [self deleteAllLocalChanges];
+            [self.userData setObject:[NSDate date] forKey:@"lastUpdated"];
+        }
+    }];
+}
+
+- (void)deleteAllLocalChanges {
+    NSBatchDeleteRequest *delete = [[NSBatchDeleteRequest alloc] initWithFetchRequest:LocalChange.fetchRequest];
+    NSPersistentStoreCoordinator *persistentStoreCoordinator = ((AppDelegate *)UIApplication.sharedApplication.delegate).persistentContainer.persistentStoreCoordinator;
+    [persistentStoreCoordinator executeRequest:delete withContext:self.context error:nil];
+}
+
+- (void)syncLocalChanges:(NSArray<LocalChange *> *)localChanges {
+    for (LocalChange *change in localChanges) {
+        [self syncEventToParse:change.oldEvent updatedEvent:change.updatedEvent];
+    }
+}
+
+- (void)didChangeOffline {
+    self.isSynced = false;
+}
+
+- (void)syncEventToParse:(Event *)oldEvent
+            updatedEvent:(Event *)newEvent {
+    if (!oldEvent) {
+        [self syncNewEventToParse:newEvent];
+    } else if (!newEvent) {
+        [self syncDeleteToParse:oldEvent];
+    } else {
+        [self syncUpdateToParse:newEvent];
+    }
+}
+
+- (void)didChangeEvent:(Event *)oldEvent
+          updatedEvent:(Event *)newEvent {
+    if (self.networkHandler.isOnline) {
+        [self syncEventToParse:oldEvent updatedEvent:newEvent];
+    } else {
+        [self saveNewLocalChange:oldEvent updatedEvent:newEvent];
+    }
+}
+
+- (void)saveNewLocalChange:(Event *)oldEvent
+              updatedEvent:(Event *)newEvent {
+    LocalChange *localChange = [[LocalChange alloc] initWithContext:self.context];
+    localChange.timestamp = [NSDate date];
+    if (oldEvent) {
+        localChange.oldEvent = [[Event alloc] initWithOriginalEvent:oldEvent];
+    }
+    if (newEvent) {
+        localChange.updatedEvent = [[Event alloc] initWithOriginalEvent:newEvent];
+    }
+    [self.context save:nil];
```

Comment:
I would think about ways to test this. 

Okay strategy would be to extract the logic for the local change into a more testable method that takes the same inputs but returns the local change. Then this method would call that and save the context.

Better strategy would be to abstract and inject a context that can be spied on for tests.